KIPs/kip-146.md

```diff
@@ -0,0 +1,169 @@
+
+Meanwhile, a few malicious validators must not be able to forge the randomness to arbitrarily determine the upcoming proposers. If not, it may increase the risk of censorship or monopolizing block rewards. Therefore, the new random proposer selection policy should be both unpredictable and verifiable.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### Parameters
+
+| Constant     | Value |
+| ------------ | ----- |
+| `FORK_BLOCK` | TBD   |
+
+### Shuffling
+
+A shuffling algorithm must be defined which will act as a building block of proposer and committee selection.
+
+[Fisher-Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle) must be used in the below pseudocode (i.e., `random.shuffle`), and [Golang's RNG](https://cs.opensource.google/go/go/+/master:src/math/rand/rng.go;l=7-12) must be used as a pseudo random number generator internally in Fisher-Yates.
+
+```py
+def shuffle_validators_KIP146(validators: list[int], round: int, seed: int):
+    """returns a list of randomly shuffled validators.
+
+    Keyword arguments:
+    validators -- deterministically sorted list of validators.
+    round -- the target consensus round.
+    seed -- `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer. The `mixHash` of the previous block is used (see KIP-114).
+    """
+    tmp = validators[:]
+    ret = []
+
+    random.seed(seed)
+    random.shuffle(tmp)
+    ret.append(tmp[0])
+
+    for i in range(round):
+        tmp = tmp[i + 1 :]
+        if len(tmp) == 0:
+            break
+        random.shuffle(tmp)
+        ret.append(tmp[0])
+    return ret
+```
+
+### Proposer selection
+
+For calculating the proposer of `FORK_BLOCK` and afterward, the new logic must be used.
+This only affects the chain whose proposer selection policy is set to `WeightedRandom`.
+
+This design allows a certain validator to be selected as the proposer for consecutive blocks, while guaranteeing another proposer to be selected in case of round change.
+
+```py
+def proposer_selector(validators, round, seed):
+    """returns a proposer from validators.
+
+    Keyword arguments:
+    validators -- deterministically sorted list of validators.
+    round -- the target consensus round.
+    seed -- `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer. The `mixHash` of the previous block is used (see KIP-114).
+    """
+    return shuffle_validators_KIP146(validators, round, seed)[round % len(validators)]
+```
+
+### Committee selection
+
+For calculating the committee of `FORK_BLOCK` and afterward, the new logic must be used.
+This only affects the chain whose proposer selection policy is set to `WeightedRandom`.
+
+This design guarantees that the given proposer address is always in the committee.
```

Comment:
The current implementation of Klaytn always designates a proposer to the committee, which appears to be a backward compatibility measure. However, in my opinion, it would be more logical for the proposer not to be included in the committee list, since the proposer's answer is always considered true. If, for instance, an attacker were to become a proposer, this would lead to the omission of verification by an honest validator.

If I have misunderstood anything, kindly correct me. Thank you.

Reply:
Your suggestion offers a valid point, I have to disagree. Removing the proposer from the committee would effectively lower the count of validators that can be fault-tolerant while maintaining consensus. This is a consequence that I believe would not serve our best interests.

Reply:
I couldn't derive any insights from your previous comment. Could you please highlight what is incorrect in my understanding?

If a malicious proposer is part of the committee, his response should be deemed invalid and disregarded. However, if a proposer is not part of the committee, then the exclusion of his response as invalid can be reconsidered. This was the basis of my inquiry.

Reply:
Even if the proposer is part of the committee, its response is not always accepted as valid by default. It still undergoes the same verification process as the rest of the validators in the committee. 
In terms of fault tolerance, if a proposer is part of the committee, they contribute to the total number of validators who can be faulty (for example, due to a malicious act or system failure) while still maintaining consensus. If the proposer is removed from the committee, then the maximum number of validators who can be faulty while still reaching a consensus will decrease by one. This could potentially affect the robustness of the network.

Reply:
Let's say there are 4 nodes (1 proposer and 3 committee members), the algorithm can tolerate 1 faulty node.
However, if you remove the proposer from the committee, leaving 3 nodes, the algorithm can tolerate 0 faulty nodes because (3-1)/3 equals 0. So if any of the 3 nodes becomes faulty, it will not be able to reach consensus, thus decreasing the fault tolerance of the system. This is why removing the proposer from the committee could decrease fault tolerance and thus make the system less robust.

Reply:
Thank you. I understand your point now.

Your concern is primarily about the possibility of accidents occurring at any time. For example, in a scenario with 10 validators and 7 committee members, and 1 proposer, a faulty node can be present in any of the 6 committee members (one is subtracted because of the proposer). Additionally, two faulty nodes cannot be tolerated. Furthermore, I've realized another aspect: the committee size of Cypress is actually not equal to the "committee size" itself, but rather its value is "committee size - 1".

In conclusion, regarding the trade-off between including or excluding the proposer in the committee. If we exclude it, we get the followings:
(1) + Removing the proposer from a selected committee saves one wasted counting.
(2) - Excluding the proposer reduces the probability of achieving minor robustness in consensus (this probability can be defined as the "probability of faulty" divided by the "length of the committee").

Based on the statements above, I am inclined to support (1) for logical consistency, but I am not suggesting a change to be proposed. This is merely to align the logical correction.

Reply:
I appreciate your detailed response. 
I'm a bit unclear about one point you made when you mentioned that the committee size of Cypress is 'committee size - 1'. 
I'm assuming that this is because you've removed the proposer from the committee in your calculation, am I correct? Could you please elaborate on this?

Reply:
Also, I'd like to share that we are considering making the number of validators equal to the committee size. This change is being considered due to potential security vulnerabilities when the committee size is smaller than the total number of validators. 
In our moving to a permissionless setting, Balancing network security and efficiency is a complex task and we are constantly exploring ways to optimize this.

Reply:
Say, there are total 10 validators.

The current implementation gaurantees to compose committee as `{p, [p, v1..v6]}`, where `p` and `v` are proposer and validator, respectively.

Among the collected committed seals by committee, one answer by `p` is always assumed as true. Then, we can say that there are valid 6 seals, not 7 seals. Thus, it derives me that thought the committee size can be interpreted as 'committee size - 1'

> Also, I'd like to share that we are considering making the number of validators equal to the committee size. This change is being considered due to potential security vulnerabilities when the committee size is smaller than the total number of validators.

This circumstance is exactly the same. Say there are total 10 validators, then committee will be composed as:
`c = {p, [p, v2..v9]}` and `len(c) = 10`. As one seal is always true, then only 9 seals are legitimate.

Reply:
The proposer's seal is also counted as a valid seal, assuming the proposer is honest. 
So, in a committee of 10 validators, including the proposer, there should be 10 valid seals - not 9. 
The validity of the proposer's seal, like any other, depends on their honesty. Consequently, if the proposer is honest, we wwill have 10 valid seals, not just 9.

Reply:
I'm guessing you might consider this situation. Attacker (proposer) can take one voting to be considered in 2/3+ aggregation.

However, I assured that this case would not be the one you considered because my claim (release a proposer from committee) would increase the security more. As the honest validator may have a voting power (10 honest / 0 attacker), but the former (current) one always take the attacker's voting. (9 honest / 1 attacker).

If this was not in your mind, could you provide an example of a situation where a proposer disagrees with their own proposal? (i.e., a situation that proposer disagrees his/her own proposal)

Reply:
If you have anything to discuss further, please use #146